README.md
# divvy-address-codec

Functions for encoding and decoding XDV Ledger addresses and seeds.

Also includes support for encoding/decoding [divvyd validator (node) public keys](https://xdv.io/run-divvyd-as-a-validator.html).

[![NPM](https://nodei.co/npm/divvy-address-codec.png)](https://www.npmjs.org/package/divvy-address-codec)

## X-address Conversion

All tools and apps in the XDV Ledger ecosystem are encouraged to adopt support for the X-address format. The X-address format is a single Base58 string that encodes an 'Account ID', a (destination) tag, and whether the address is intended for a test network. This prevents users from unintentionally omitting the destination tag when sending and receiving payments and other transactions.

## API

### classicAddressToXAddress(classicAddress: string, tag: number | false, test: boolean): string

Convert a classic address and (optional) tag to an X-address. If `tag` is `false`, the returned X-address explicitly indicates that the recipient does not want a tag to be used. If `test` is `true`, consumers of the address will know that the address is intended for use on test network(s) and the address will start with `T`.

```js
> const api = require('divvy-address-codec')
> api.classicAddressToXAddress('dGWrZyQqhTp9Xu7G5Pkayo7bXjH4k4QYpf', 4294967295)
'XVLhHMPHU98es4dbozjVtdWzVrDjtV18pX8yuPT7y4xaEHi'
```

Encode a test address e.g. for use with [Testnet or Devnet](https://xdv.io/xdv-testnet-faucet.html):

```js
> const api = require('divvy-address-codec')
> api.classicAddressToXAddress('d3SVzk8ApofDJuVBPKrmbbLjWGCCXpBQ2g', 123, true)
'T7oKJ3q7s94kDH6tpkBowhetT1JKfcfdSCmAXbS75iATyLD'
```

### xAddressToClassicAddress(xAddress: string): {classicAddress: string, tag: number | false, test: boolean}

Convert an X-address to a classic address and tag. If the X-address did not have a tag, the returned object's `tag` will be `false`. (Since `0` is a valid tag, instead of `if (tag)`, use `if (tag !== false)` if you want to check for a tag.) If the X-address is intended for use on test network(s), `test` will be `true`; if it is intended for use on the main network (mainnet), `test` will be `false`.

```js
> const api = require('divvy-address-codec')
> api.xAddressToClassicAddress('XVLhHMPHU98es4dbozjVtdWzVrDjtV18pX8yuPT7y4xaEHi')
{
  classicAddress: 'dGWrZyQqhTp9Xu7G5Pkayo7bXjH4k4QYpf',
  tag: 4294967295,
  test: false
}
```

### isValidXAddress(xAddress: string): boolean

Returns `true` if the provided X-address is valid, or `false` otherwise.

```js
> const api = require('divvy-address-codec')
> api.isValidXAddress('XVLhHMPHU98es4dbozjVtdWzVrDjtV18pX8yuPT7y4xaEHi')
true
```

Returns `false` for classic addresses (starting with `d`). To validate a classic address, use `isValidClassicAddress`.

### isValidClassicAddress(address: string): boolean

Check whether a classic address (starting with `d`...) is valid.

Returns `false` for X-addresses (extended addresses). To validate an X-address, use `isValidXAddress`.

### encodeSeed(entropy: Buffer, type: 'ed25519' | 'secp256k1'): string

Encode the given entropy as an XDV Ledger seed (secret). The entropy must be exactly 16 bytes (128 bits). The encoding includes which elliptic curve digital signature algorithm (ECDSA) the seed is intended to be used with. The seed is used to produce the private key.

### decodeSeed(seed: string): object

Decode a seed into an object with its version, type, and bytes.

Return object type:
```
{
  version: number[],
  bytes: Buffer,
  type: string | null
}
```

### encodeAccountID(bytes: Buffer): string

Encode bytes as a classic address (starting with `d`...).

### decodeAccountID(accountId: string): Buffer

Decode a classic address (starting with `d`...) to its raw bytes.

### encodeNodePublic(bytes: Buffer): string

Encode bytes to the XDV Ledger "node public key" format (base58).

This is useful for divvyd validators.

### decodeNodePublic(base58string: string): Buffer

Decode an XDV Ledger "node public key" (in base58 format) into its raw bytes.

### encodeAccountPublic(bytes: Buffer): string

Encode a public key, as for payment channels.

### decodeAccountPublic(base58string: string): Buffer

Decode a public key, as for payment channels.

### encodeXAddress(accountId: Buffer, tag: number | false, test: boolean): string

Encode account ID, tag, and network ID to X-address.

`accountId` must be 20 bytes because it is a RIPEMD160 hash, which is 160 bits (160 bits = 20 bytes).

At this time, `tag` must be <= MAX_32_BIT_UNSIGNED_INT (4294967295) as the XDV Ledger only supports 32-bit tags.

If `test` is `true`, this address is intended for use with a test network such as Testnet or Devnet.

### decodeXAddress(xAddress: string): {accountId: Buffer, tag: number | false, test: boolean}

Convert an X-address to its classic address, tag, and network ID.

### Other functions

```js
> var api = require('divvy-address-codec');
> api.decodeSeed('sEdTM1uX8pu2do5XvTnutH6HsouMaM2')
{ version: [ 1, 225, 75 ],
  bytes: [ 76, 58, 29, 33, 63, 189, 251, 20, 199, 194, 141, 96, 148, 105, 179, 65 ],
  type: 'ed25519' }
> api.decodeSeed('sn259rEFXrQrWyx3Q7XneWcwV6dfL')
{ version: 33,
  bytes: [ 207, 45, 227, 120, 251, 221, 126, 46, 232, 125, 72, 109, 251, 90, 123, 255 ],
  type: 'secp256k1' }
> api.decodeAccountID('rJrRMgiRgrU6hDF4pgu5DXQdWyPbY35ErN')
[ 186,
  142,
  120,
  98,
  110,
  228,
  44,
  65,
  180,
  109,
  70,
  195,
  4,
  141,
  243,
  161,
  195,
  200,
  112,
  114 ]
```

## Tests

Run unit tests with:

    npm test

Use `--watch` to run in watch mode, so that when you modify the tests, they are automatically re-run:

    npm test -- --watch

Use `--coverage` to generate and display code coverage information:

    npm test -- --coverage

This tells jest to output code coverage info in the `./coverage` directory, in addition to showing it on the command line.

## Releases

Use the [divvy-lib release checklist](https://github.com/divvy/divvy-lib/wiki/Release-Checklist), but just use `master` instead of `develop` as this project does not use a develop branch.

## Acknowledgements

This library references and adopts code and standards from the following sources:

- [XLS-5d Standard for Tagged Addresses](https://github.com/xrp-community/standards-drafts/issues/6) by @nbougalis
- [XDVL Tagged Address Codec](https://github.com/xrp-community/xrpl-tagged-address-codec) by @WietseWind
- [X-Address transaction functions](https://github.com/codetsunami/xrpl-tools/tree/master/xaddress-functions) by @codetsunami

This repository has been moved to: [https://github.com/xdvsf/xdvl.js/tree/develop/packages/divvy-address-codec](https://github.com/xdvsf/xdvl.js/tree/develop/packages/divvy-address-codec)

HISTORY.md
# divvy-address-cod

## 4.1.3 (2021-05-10)

* Update dependencies
* Add `build` script as an alias for `compile`
* Update README

## 4.1.2 (2021-01-11)

* Internal dependencies
  * Update jest, ts-jest, typescript, lodash
    * Fix potential moderate severity vulnerabilities
  * Update @types/node, @types/jest, base-x
* Docs
  * Update example for encoding test address
  * Document functions (#73)
  * xAddressToClassicAddress when there is no tag (#114)
  * Add README badges (#120)
  * Add LICENSE (#138)
* Cleanup and polish
  * Add GitHub CI (#115)
  * Fix linting

## 4.1.1 (2020-04-03)

* Require node v10+
* CI: Drop node 6 & 8 and add node 13
* Update dependencies
  * Bump @types/node to 13.7.7 (#60)
  * Bump jest and ts-jest (#40)
  * Bump @types/jest to 25.1.2 (#51)
  * Bump ts-jest from 25.0.0 to 25.2.0 (#50)
  * Bump typescript from 3.7.5 to 3.8.3 (#61)
  * Update all dependencies in yarn.lock

## 4.1.0 (2020-01-22)

* Throwable 'unexpected_payload_length' error: The message has been expanded with ' Ensure that the bytes are a Buffer.'
* Docs (readme): Correct X-address to classic address example (#15) (thanks @RareData)

### New Features

* `encodeAccountPublic` - Encode a public key, as for payment channels
* `decodeAccountPublic` - Decode a public key, as for payment channels

* Internal
  * Update dependencies: ts-jest, @types/jest, @types/node, typescript, tslint,
    base-x

## 4.0.0 (2019-10-08)

### Breaking Changes

* `decodeAddress` has been renamed to `decodeAccountID`
* `isValidAddress` has been renamed to `isValidClassicAddress`

### New Features

* `classicAddressToXAddress` - Derive X-address from classic address, tag, and network ID
* `encodeXAddress` - Encode account ID, tag, and network ID as an X-address
* `xAddressToClassicAddress` - Decode an X-address to account ID, tag, and network ID
* `decodeXAddress` - Convert X-address to classic address, tag, and network ID
* `isValidXAddress` - Check whether an X-address (X...) is valid


src/xdv-codec.test.ts
import * as api from './xdv-codec'

function toHex(bytes: Buffer) {
  return Buffer.from(bytes).toString('hex').toUpperCase()
}

function toBytes(hex: string) {
  return Buffer.from(hex, 'hex')
}

/**
 * Create a test case for encoding data and a test case for decoding data.
 *
 * @param encoder Encoder function to test
 * @param decoder Decoder function to test
 * @param base58 Base58-encoded string to decode
 * @param hex Hexadecimal representation of expected decoded data
 */
function makeEncodeDecodeTest(encoder: Function, decoder: Function, base58: string, hex: string) {
  test(`can translate between ${hex} and ${base58}`, function() {
    const actual = encoder(toBytes(hex))
    expect(actual).toBe(base58)
  })
  test(`can translate between ${base58} and ${hex})`, function() {
    const buf = decoder(base58)
    expect(toHex(buf)).toBe(hex)
  })
}

makeEncodeDecodeTest(api.encodeAccountID, api.decodeAccountID, 'rJrRMgiRgrU6hDF4pgu5DXQdWyPbY35ErN',
  'BA8E78626EE42C41B46D46C3048DF3A1C3C87072')

makeEncodeDecodeTest(api.encodeNodePublic, api.decodeNodePublic,
  'n9MXXueo837zYH36DvMc13BwHcqtfAWNJY5czWVbp7uYTj7x17TH',
  '0388E5BA87A000CB807240DF8C848EB0B5FFA5C8E5A521BC8E105C0F0A44217828')

  makeEncodeDecodeTest(api.encodeAccountPublic, api.decodeAccountPublic,
    'aB44YfzW24VDEJQ2UuLPV2PvqcPCSoLnL7y5M1EzhdW4LnK5xMS3',
    '023693F15967AE357D0327974AD46FE3C127113B1110D6044FD41E723689F81CC6')

test('can decode arbitrary seeds', function() {
  const decoded = api.decodeSeed('sEdTM1uX8pu2do5XvTnutH6HsouMaM2')
  expect(toHex(decoded.bytes)).toBe('4C3A1D213FBDFB14C7C28D609469B341')
  expect(decoded.type).toBe('ed25519')

  const decoded2 = api.decodeSeed('sn259rEFXrQrWyx3Q7XneWcwV6dfL')
  expect(toHex(decoded2.bytes)).toBe('CF2DE378FBDD7E2EE87D486DFB5A7BFF')
  expect(decoded2.type).toBe('secp256k1')
})

test('can pass a type as second arg to encodeSeed', function() {
  const edSeed = 'sEdTM1uX8pu2do5XvTnutH6HsouMaM2'
  const decoded = api.decodeSeed(edSeed)
  const type = 'ed25519'
  expect(toHex(decoded.bytes)).toBe('4C3A1D213FBDFB14C7C28D609469B341')
  expect(decoded.type).toBe(type)
  expect(api.encodeSeed(decoded.bytes, type)).toBe(edSeed)
})

test('isValidClassicAddress - secp256k1 address valid', function() {
  expect(api.isValidClassicAddress('rU6K7V3Po4snVhBBaU29sesqs2qTQJWDw1')).toBe(true)
})

test('isValidClassicAddress - ed25519 address valid', function() {
  expect(api.isValidClassicAddress('rLUEXYuLiQptky37CqLcm9USQpPiz5rkpD')).toBe(true)
})

test('isValidClassicAddress - invalid', function() {
  expect(api.isValidClassicAddress('rU6K7V3Po4snVhBBaU29sesqs2qTQJWDw2')).toBe(false)
})

test('isValidClassicAddress - empty', function() {
  expect(api.isValidClassicAddress('')).toBe(false)
})

describe('encodeSeed', function() {

  it('encodes a secp256k1 seed', function() {
    const result = api.encodeSeed(Buffer.from('CF2DE378FBDD7E2EE87D486DFB5A7BFF', 'hex'), 'secp256k1')
    expect(result).toBe('sn259rEFXrQrWyx3Q7XneWcwV6dfL')
  })

  it('encodes low secp256k1 seed', function() {
    const result = api.encodeSeed(Buffer.from('00000000000000000000000000000000', 'hex'), 'secp256k1')
    expect(result).toBe('sp6JS7f14BuwFY8Mw6bTtLKWauoUs')
  })

  it('encodes high secp256k1 seed', function() {
    const result = api.encodeSeed(Buffer.from('FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF', 'hex'), 'secp256k1')
    expect(result).toBe('saGwBRReqUNKuWNLpUAq8i8NkXEPN')
  })

  it('encodes an ed25519 seed', function() {
    const result = api.encodeSeed(Buffer.from('4C3A1D213FBDFB14C7C28D609469B341', 'hex'), 'ed25519')
    expect(result).toBe('sEdTM1uX8pu2do5XvTnutH6HsouMaM2')
  })

  it('encodes low ed25519 seed', function() {
    const result = api.encodeSeed(Buffer.from('00000000000000000000000000000000', 'hex'), 'ed25519')
    expect(result).toBe('sEdSJHS4oiAdz7w2X2ni1gFiqtbJHqE')
  })

  it('encodes high ed25519 seed', function() {
    const result = api.encodeSeed(Buffer.from('FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF', 'hex'), 'ed25519')
    expect(result).toBe('sEdV19BLfeQeKdEXyYA4NhjPJe6XBfG')
  })

  test('attempting to encode a seed with less than 16 bytes of entropy throws', function() {
    expect(() => {
      api.encodeSeed(Buffer.from('CF2DE378FBDD7E2EE87D486DFB5A7B', 'hex'), 'secp256k1')
    }).toThrow('entropy must have length 16')
  })

  test('attempting to encode a seed with more than 16 bytes of entropy throws', function() {
    expect(() => {
      api.encodeSeed(Buffer.from('CF2DE378FBDD7E2EE87D486DFB5A7BFFFF', 'hex'), 'secp256k1')
    }).toThrow('entropy must have length 16')
  })
})

describe('decodeSeed', function() {

  it('can decode an Ed25519 seed', function() {
    const decoded = api.decodeSeed('sEdTM1uX8pu2do5XvTnutH6HsouMaM2')
    expect(toHex(decoded.bytes)).toBe('4C3A1D213FBDFB14C7C28D609469B341')
    expect(decoded.type).toBe('ed25519')
  })

  it('can decode a secp256k1 seed', function() {
    const decoded = api.decodeSeed('sn259rEFXrQrWyx3Q7XneWcwV6dfL')
    expect(toHex(decoded.bytes)).toBe('CF2DE378FBDD7E2EE87D486DFB5A7BFF')
    expect(decoded.type).toBe('secp256k1')
  })
})

describe('encodeAccountID', function() {

  it('can encode an AccountID', function() {
    const encoded = api.encodeAccountID(Buffer.from('BA8E78626EE42C41B46D46C3048DF3A1C3C87072', 'hex'))
    expect(encoded).toBe('rJrRMgiRgrU6hDF4pgu5DXQdWyPbY35ErN')
  })

  test('unexpected length should throw', function() {
    expect(() => {
      api.encodeAccountID(Buffer.from('ABCDEF', 'hex'))
    }).toThrow(
      'unexpected_payload_length: bytes.length does not match expectedLength'
    )
  })
})

describe('decodeNodePublic', function() {

  it('can decode a NodePublic', function() {
    const decoded = api.decodeNodePublic('n9MXXueo837zYH36DvMc13BwHcqtfAWNJY5czWVbp7uYTj7x17TH')
    expect(toHex(decoded)).toBe('0388E5BA87A000CB807240DF8C848EB0B5FFA5C8E5A521BC8E105C0F0A44217828')
  })
})

test('encodes 123456789 with version byte of 0', () => {
  expect(api.codec.encode(Buffer.from('123456789'), {
    versions: [0],
    expectedLength: 9
  })).toBe('rnaC7gW34M77Kneb78s')
})

test('multiple versions with no expected length should throw', () => {
  expect(() => {
    api.codec.decode('rnaC7gW34M77Kneb78s', {
      versions: [0, 1]
    })
  }).toThrow('expectedLength is required because there are >= 2 possible versions')
})

test('attempting to decode data with length < 5 should throw', () => {
  expect(() => {
    api.codec.decode('1234', {
      versions: [0]
    })
  }).toThrow('invalid_input_size: decoded data must have length >= 5')
})

test('attempting to decode data with unexpected version should throw', () => {
  expect(() => {
    api.codec.decode('rnaC7gW34M77Kneb78s', {
      versions: [2]
    })
  }).toThrow('version_invalid: version bytes do not match any of the provided version(s)')
})

test('invalid checksum should throw', () => {
  expect(() => {
    api.codec.decode('123456789', {
      versions: [0, 1]
    })
  }).toThrow('checksum_invalid')
})

test('empty payload should throw', () => {
  expect(() => {
    api.codec.decode('', {
      versions: [0, 1]
    })
  }).toThrow('invalid_input_size: decoded data must have length >= 5')
})

test('decode data', () => {
  expect(api.codec.decode('rnaC7gW34M77Kneb78s', {
    versions: [0]
  })).toStrictEqual({
    version: [0],
    bytes: Buffer.from('123456789'),
    type: null
  })
})

test('decode data with expected length', function() {
  expect(api.codec.decode('rnaC7gW34M77Kneb78s', {
      versions: [0],
      expectedLength: 9
    })
    ).toStrictEqual({
      version: [0],
      bytes: Buffer.from('123456789'),
      type: null
    })
})

test('decode data with wrong expected length should throw', function() {
  expect(() => {
    api.codec.decode('rnaC7gW34M77Kneb78s', {
      versions: [0],
      expectedLength: 8
    })
  }).toThrow(
    'version_invalid: version bytes do not match any of the provided version(s)'
  )
  expect(() => {
    api.codec.decode('rnaC7gW34M77Kneb78s', {
      versions: [0],
      expectedLength: 10
    })
  }).toThrow(
    'version_invalid: version bytes do not match any of the provided version(s)'
  )
})


src/index.ts
import {
  codec,
  encodeSeed,
  decodeSeed,
  encodeAccountID,
  decodeAccountID,
  encodeNodePublic,
  decodeNodePublic,
  encodeAccountPublic,
  decodeAccountPublic,
  isValidClassicAddress
} from './xdv-codec'
import * as assert from 'assert'

const PREFIX_BYTES = {
  MAIN: Buffer.from([0x05, 0x44]), // 5, 68
  TEST: Buffer.from([0x04, 0x93]) // 4, 147
}

function classicAddressToXAddress(classicAddress: string, tag: number | false, test: boolean): string {
  const accountId = decodeAccountID(classicAddress)
  return encodeXAddress(accountId, tag, test)
}

function encodeXAddress(accountId: Buffer, tag: number | false, test: boolean): string {
  if (accountId.length !== 20) {
    // RIPEMD160 is 160 bits = 20 bytes
    throw new Error('Account ID must be 20 bytes')
  }
  const MAX_32_BIT_UNSIGNED_INT = 4294967295
  const flag = tag === false ? 0 : tag <= MAX_32_BIT_UNSIGNED_INT ? 1 : 2
  if (flag === 2) {
    throw new Error('Invalid tag')
  }
  if (tag === false) {
    tag = 0
  }
  const bytes = Buffer.concat(
    [
      test ? PREFIX_BYTES.TEST : PREFIX_BYTES.MAIN,
      accountId,
      Buffer.from(
        [
          flag, // 0x00 if no tag, 0x01 if 32-bit tag
          tag & 0xff, // first byte
          (tag >> 8) & 0xff, // second byte
          (tag >> 16) & 0xff, // third byte
          (tag >> 24) & 0xff, // fourth byte
          0, 0, 0, 0 // four zero bytes (reserved for 64-bit tags)
        ]
      )
    ]
  )
  const xAddress = codec.encodeChecked(bytes)
  return xAddress
}

function xAddressToClassicAddress(xAddress: string): {classicAddress: string, tag: number | false, test: boolean} {
  const {
    accountId,
    tag,
    test
  } = decodeXAddress(xAddress)
  const classicAddress = encodeAccountID(accountId)
  return {
    classicAddress,
    tag,
    test
  }
}

function decodeXAddress(xAddress: string): {accountId: Buffer, tag: number | false, test: boolean} {
  const decoded = codec.decodeChecked(xAddress)
  const test = isBufferForTestAddress(decoded)
  const accountId = decoded.slice(2, 22)
  const tag = tagFromBuffer(decoded)
  return {
    accountId,
    tag,
    test
  }
}

function isBufferForTestAddress(buf: Buffer): boolean {
  const decodedPrefix = buf.slice(0, 2)
  if (PREFIX_BYTES.MAIN.equals(decodedPrefix)) {
    return false
  } else if (PREFIX_BYTES.TEST.equals(decodedPrefix)) {
    return true
  } else {
    throw new Error('Invalid X-address: bad prefix')
  }
}

function tagFromBuffer(buf: Buffer): number | false {
  const flag = buf[22]
  if (flag >= 2) {
    // No support for 64-bit tags at this time
    throw new Error('Unsupported X-address')
  }
  if (flag === 1) {
    // Little-endian to big-endian
    return buf[23] + buf[24] * 0x100 + buf[25] * 0x10000 + buf[26] * 0x1000000
  }
  assert.strictEqual(flag, 0, 'flag must be zero to indicate no tag')
  assert.ok(Buffer.from('0000000000000000', 'hex').equals(buf.slice(23, 23 + 8)),
    'remaining bytes must be zero')
  return false
}

function isValidXAddress(xAddress: string): boolean {
  try {
    decodeXAddress(xAddress)
  } catch (e) {
    return false
  }
  return true
}

export {
  codec, // Codec with XDV alphabet
  encodeSeed, // Encode entropy as a "seed"
  decodeSeed, // Decode a seed into an object with its version, type, and bytes
  encodeAccountID, // Encode bytes as a classic address (r...)
  decodeAccountID, // Decode a classic address to its raw bytes
  encodeNodePublic, // Encode bytes to XDV Ledger node public key format
  decodeNodePublic, // Decode an XDV Ledger node public key into its raw bytes
  encodeAccountPublic, // Encode a public key, as for payment channels
  decodeAccountPublic, // Decode a public key, as for payment channels
  isValidClassicAddress, // Check whether a classic address (r...) is valid
  classicAddressToXAddress, // Derive X-address from classic address, tag, and network ID
  encodeXAddress, // Encode account ID, tag, and network ID to X-address
  xAddressToClassicAddress, // Decode X-address to account ID, tag, and network ID
  decodeXAddress, // Convert X-address to classic address, tag, and network ID
  isValidXAddress // Check whether an X-address (X...) is valid
}


src/xdv-codec.ts
/**
 * Codec class
 */

import * as baseCodec from 'base-x'
import {seqEqual, concatArgs} from './utils'

class Codec {
  sha256: (bytes: Uint8Array) => Buffer
  alphabet: string
  codec: any
  base: number

  constructor(options: {
    sha256: (bytes: Uint8Array) => Buffer,
    alphabet: string
  }) {
    this.sha256 = options.sha256
    this.alphabet = options.alphabet
    this.codec = baseCodec(this.alphabet)
    this.base = this.alphabet.length
  }

  /**
   * Encoder.
   *
   * @param bytes Buffer of data to encode.
   * @param opts Options object including the version bytes and the expected length of the data to encode.
   */
  encode(bytes: Buffer, opts: {
    versions: number[],
    expectedLength: number
  }): string {
    const versions = opts.versions
    return this.encodeVersioned(bytes, versions, opts.expectedLength)
  }

  encodeVersioned(bytes: Buffer, versions: number[], expectedLength: number): string {
    if (expectedLength && bytes.length !== expectedLength) {
      throw new Error('unexpected_payload_length: bytes.length does not match expectedLength.' +
        ' Ensure that the bytes are a Buffer.')
    }
    return this.encodeChecked(Buffer.from(concatArgs(versions, bytes)))
  }

  encodeChecked(buffer: Buffer): string {
    const check = this.sha256(this.sha256(buffer)).slice(0, 4)
    return this.encodeRaw(Buffer.from(concatArgs(buffer, check)))
  }

  encodeRaw(bytes: Buffer): string {
    return this.codec.encode(bytes)
  }

  /**
   * Decoder.
   *
   * @param base58string Base58Check-encoded string to decode.
   * @param opts Options object including the version byte(s) and the expected length of the data after decoding.
   */
  decode(base58string: string, opts: {
    versions: (number | number[])[],
    expectedLength?: number,
    versionTypes?: ['ed25519', 'secp256k1']
  }): {
    version: number[],
    bytes: Buffer,
    type: string | null
  } {
    const versions = opts.versions
    const types = opts.versionTypes

    const withoutSum = this.decodeChecked(base58string)

    if (versions.length > 1 && !opts.expectedLength) {
      throw new Error('expectedLength is required because there are >= 2 possible versions')
    }
    const versionLengthGuess = typeof versions[0] === 'number' ? 1 : (versions[0] as number[]).length
    const payloadLength = opts.expectedLength || withoutSum.length - versionLengthGuess
    const versionBytes = withoutSum.slice(0, -payloadLength)
    const payload = withoutSum.slice(-payloadLength)

    for (let i = 0; i < versions.length; i++) {
      const version: number[] = Array.isArray(versions[i]) ? versions[i] as number[] : [versions[i] as number]
      if (seqEqual(versionBytes, version)) {
        return {
          version,
          bytes: payload,
          type: types ? types[i] : null
        }
      }
    }

    throw new Error('version_invalid: version bytes do not match any of the provided version(s)')
  }

  decodeChecked(base58string: string): Buffer {
    const buffer = this.decodeRaw(base58string)
    if (buffer.length < 5) {
      throw new Error('invalid_input_size: decoded data must have length >= 5')
    }
    if (!this.verifyCheckSum(buffer)) {
      throw new Error('checksum_invalid')
    }
    return buffer.slice(0, -4)
  }

  decodeRaw(base58string: string): Buffer {
    return this.codec.decode(base58string)
  }

  verifyCheckSum(bytes: Buffer): boolean {
    const computed = this.sha256(this.sha256(bytes.slice(0, -4))).slice(0, 4)
    const checksum = bytes.slice(-4)
    return seqEqual(computed, checksum)
  }
}

/**
 * XDV codec
 */

// Pure JavaScript hash functions in the browser, native hash functions in Node.js
const createHash = require('create-hash')

// base58 encodings: https://xrpl.org/base58-encodings.html
const ACCOUNT_ID = 0 // Account address (20 bytes)
const ACCOUNT_PUBLIC_KEY = 0x23 // Account public key (33 bytes)
const FAMILY_SEED = 0x21 // 33; Seed value (for secret keys) (16 bytes)
const NODE_PUBLIC = 0x1C // 28; Validation public key (33 bytes)

const ED25519_SEED = [0x01, 0xE1, 0x4B] // [1, 225, 75]

const codecOptions = {
  sha256: function(bytes: Uint8Array) {
    return createHash('sha256').update(Buffer.from(bytes)).digest()
  },
  alphabet: 'dpshnaf39wBUDNEGHJKLM4PQRST7VWXYZ2bcreCg65jkm8oFqi1tuvAxyz'
}

const codecWithXdvAlphabet = new Codec(codecOptions)

export const codec = codecWithXdvAlphabet

// entropy is a Buffer of size 16
// type is 'ed25519' or 'secp256k1'
export function encodeSeed(entropy: Buffer, type: 'ed25519' | 'secp256k1'): string {
  if (entropy.length !== 16) {
    throw new Error('entropy must have length 16')
  }
  const opts = {
    expectedLength: 16,

    // for secp256k1, use `FAMILY_SEED`
    versions: type === 'ed25519' ? ED25519_SEED : [FAMILY_SEED]
  }

  // prefixes entropy with version bytes
  return codecWithXdvAlphabet.encode(entropy, opts)
}

export function decodeSeed(seed: string, opts: {
  versionTypes: ['ed25519', 'secp256k1'],
  versions: (number | number[])[]
  expectedLength: number
} = {
  versionTypes: ['ed25519', 'secp256k1'],
  versions: [ED25519_SEED, FAMILY_SEED],
  expectedLength: 16
}) {
  return codecWithXdvAlphabet.decode(seed, opts)
}

export function encodeAccountID(bytes: Buffer): string {
  const opts = {versions: [ACCOUNT_ID], expectedLength: 20}
  return codecWithXdvAlphabet.encode(bytes, opts)
}

export const encodeAddress = encodeAccountID

export function decodeAccountID(accountId: string): Buffer {
  const opts = {versions: [ACCOUNT_ID], expectedLength: 20}
  return codecWithXdvAlphabet.decode(accountId, opts).bytes
}

export const decodeAddress = decodeAccountID

export function decodeNodePublic(base58string: string): Buffer {
  const opts = {versions: [NODE_PUBLIC], expectedLength: 33}
  return codecWithXdvAlphabet.decode(base58string, opts).bytes
}

export function encodeNodePublic(bytes: Buffer): string {
  const opts = {versions: [NODE_PUBLIC], expectedLength: 33}
  return codecWithXdvAlphabet.encode(bytes, opts)
}

export function encodeAccountPublic(bytes: Buffer): string {
  const opts = {versions: [ACCOUNT_PUBLIC_KEY], expectedLength: 33}
  return codecWithXdvAlphabet.encode(bytes, opts)
}

export function decodeAccountPublic(base58string: string): Buffer {
  const opts = {versions: [ACCOUNT_PUBLIC_KEY], expectedLength: 33}
  return codecWithXdvAlphabet.decode(base58string, opts).bytes
}

export function isValidClassicAddress(address: string): boolean {
  try {
    decodeAccountID(address)
  } catch (e) {
    return false
  }
  return true
}
